File: src/converter/plugins/convertImportType.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export const convertImportType = createSimplePlugin((node, context, render) => {
    if (!ts.isImportTypeNode(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    checkCoverageService?.cover(node)
    checkCoverageService?.deepCover(node.argument)

    const qualifier = node.qualifier
        ? render(node.qualifier)
        : "Any"

    return `/* import(${typeScriptService?.printNode(node.argument)}) */ ${qualifier}`
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertImportType.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isImportTypeNode

val convertImportType = createPlugin plugin@{ node, context, render ->
    if (!isImportTypeNode(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    checkCoverageService?.cover(node)
    checkCoverageService?.deepCover(node.argument)

    val qualifier = node.qualifier?.let { render(it) } ?: "Any"

    "/* import(${typeScriptService?.printNode(node.argument)}) */ $qualifier"
}
